### PR TITLE
Search addition

### PR DESCRIPTION
**Pull request by usselite**
Implements the SEARCH command.

### CODE REVIEW

File: src/WebDav.Client/IWebDavClient.cs [hunk 1]
```diff
@@ -400,5 +401,13 @@ public interface IWebDavClient : IDisposable
...
         /// <returns>An instance of <see cref="WebDavResponse" />.</returns>
         Task<WebDavResponse> Unlock(Uri requestUri, UnlockParameters parameters);
+
+        /// <summary>
+        /// Use WebDav SEARCH Method.
+        /// </summary>
+        /// <param name="requestUri">Request uri.</param>
+        /// <param name="parameters">SEARCH parameters.</param>
+        /// <returns></returns>
+        Task<WebDavResponse> Search(Uri requestUri, SearchParameters parameters)
```
Comment (skazantsev) [inline src/WebDav.Client/IWebDavClient.cs:411]: Missing semicolon, doesn't build.

File: src/WebDav.Client/Request/SearchParameters.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Xml.Linq;
+
+namespace WebDav.Client.Request
+{
+    public class SearchParameters
```
Comment (skazantsev) [inline src/WebDav.Client/Request/SearchParameters.cs:9]: I'd recommend initializing collections like with empty lists (similar to other `XxxParameters` classes).

Comment (skazantsev) [reply src/WebDav.Client/Request/SearchParameters.cs:9]: nit: also the formatting of the class is a bit off.

File: src/WebDav.Client/Request/SearchRequestBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    internal static class SearchRequestBuilder
+    {
+        /// <summary>
+        /// Build the WEBDAV XML body.
+        /// </summary>
+        /// <param name="select">Set arguments to select parameters</param>
+        /// <param name="fromDirectory">Set root directory to search from.</param>
+        /// <param name="searchKeyWord">Set search keyword, e.g. 'test%' % = like.</param>
+        /// <param name="where">Set argument where to search for.</param>
+        /// <param name="namespaces">Optional, add namespaces.</param>
```
Comment (skazantsev) [inline src/WebDav.Client/Request/SearchRequestBuilder.cs:16]: The `namespaces` parameter is nullable and when it is `null` then the method fails with `NullReferenceException` on `foreach (var ns in namespaces)` line.

My recommendation is to pass `SearchParameters` to this method and make sure collections are at least initialized with empty lists.

File: src/WebDav.Client/WebDavClient.cs [hunk 4]
```diff
@@ -732,6 +733,28 @@ public async Task<WebDavResponse> Unlock(Uri requestUri, UnlockParameters parame
             return new WebDavResponse((int)response.StatusCode, response.ReasonPhrase);
         }
 
+        /// <summary>
+        /// Use WebDav SEARCH command.
+        /// </summary>
+        /// <param name="requestUri">Base URI.</param>
+        /// <param name="parameters">Set additional search parameters.</param>
+        /// <returns></returns>
+        public async Task<WebDavResponse> Search(Uri requestUri, SearchParameters parameters)
```
Comment (skazantsev) [inline src/WebDav.Client/WebDavClient.cs:742]: Wouldn't it be better to return `PropfindResponse` class from method?

**Review by skazantsev — COMMENTED** [review]
I've added a couple of comments. Also, is it possible to add tests for `SearchRequestBuilder`?

Thank you!

File: src/WebDav.Client/IWebDavClient.cs (hunk 1 above)
Comment (usselite) [reply src/WebDav.Client/IWebDavClient.cs]: Sorry, my mistake, copied it in a hurry between two projects.

File: src/WebDav.Client/Request/SearchParameters.cs (hunk 2 above)
Comment (usselite) [reply src/WebDav.Client/Request/SearchParameters.cs:9]: Fixed

File: src/WebDav.Client/WebDavClient.cs (hunk 4 above)
Comment (usselite) [reply src/WebDav.Client/WebDavClient.cs]: Changed it to PropFindResponse

**Comment by usselite** [comment]
I did add two tests for search request builder. I have to admit, my understanding of webdav is limited, especially to build more detailed tests like those already existent.

File: src/WebDav.Client/Request/SearchRequestBuilder.cs (hunk 3 above)
Comment (usselite) [reply src/WebDav.Client/Request/SearchRequestBuilder.cs]: Solved with initialization of all parameters.

**Comment by skazantsev** [comment]
Thanks a lot for the contribution, I'll merge it and release a new version.

**Comment by skazantsev** [comment]
[A new version](https://www.nuget.org/packages/WebDav.Client/2.8.0) is in nuget.
@usselite I also did some refactoring & minor code changes.

**Pull request merged**